ee/dynamodb/events.py

```diff
@@ -0,0 +1,38 @@
+import boto3  # type: ignore
+
+from ee.clickhouse.models.event import delete_event, update_event
+from ee.dynamodb.models.events import Event
+from posthog.settings import DYNAMODB_URL
+
+DYNAMODB_EVENTS_TABLE = "Events"
+
+
+def create_events_table(*args, **kwargs):
+    Event.create_table(read_capacity_units=1, write_capacity_units=1)
+    return
+
+
+def destroy_events_table():
+    Event.delete_table()
+
+
+def ensure_events_table(*args, **kwargs):
+    """
+    Drops and then recreates the events table ensuring it is empty
+    """
+    try:
+        destroy_events_table()
+    except:
+        pass
+
+    create_events_table()
+
+
+def update_event_person(distinct_id: str, person_uuid: str) -> None:
+    events = Event.query(distinct_id)
+    for event in events:
+        delete_event(event)
+        event.person_uuid = person_uuid
+        event.save()
+        update_event(event)
+    return
```

Comment:
This looks like a pretty clever way to solve this! Though I'm still very concerned about timing edge cases [like I wrote here](https://github.com/PostHog/posthog/issues/1929#issuecomment-713434921). With our scale it's a matter of when we run into them, not a matter of if.

Suppose in one worker `events = Event.query(distinct_id)` takes 1 sec to respond. While it's responding, another worker runs the same query with the same parameters... and and gets exactly the same list. Now both of them will write `-1` entries with the previous person_uuid and `+1` entries with the new person_uuids. The end result: garbage data that's very hard to fix after.

How to avoid this from becoming an issue?

Reply:
From [the docs](https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/collapsingmergetree/):

> The SELECT results depend strongly on the consistency of object changes history. Be accurate when preparing data for inserting.

and:

> “Cancel” string should contain copies of the sorting key fields of the “state” string and the opposite Sign.

The current events collapsing merge tree has the following sorting key:

```
ORDER BY (team_id, toDate(timestamp), distinct_id, uuid)
```

So I guess without `person_uuid` in the list, this is actually fine and I'm over thinking it. :)

Reply:
hey! sorry for lag here I didn't realize anyone was reviewing this. Just to address the first worry:

The deletes are idempotent. You have have 100 processes all trying to delete the same event here which would just push the sign deep into the negative numbers when you go to grab the row. It would also push the updated value deep into the positive integers. This happens because you are summing the sign for the combo of values you are trying to get here, in this case the sort keys + person_uuid. When you are summing it up you are filtering the newest version by saying `having sum(sign) > 0` which is kind of magical because all you need to do is have `a > b` where `a` is the number of times you emit the new value an `b` is the number of times you have emitted the new value. So the real worry here is that you have the process die on after deleting the event and before updating the event w/ the new person_uuid. If that did happen it would only effect one event as it is written which isn't too bad. In that case I would probably worry about the events that were not updated more than the single inconsistent event. We could try to make this transactional, but I feel like it could be an over optimization.